[azure_logs] Add Custom Azure Logs input package (#11552)
elastic/integrations: packages/azure_logs/docs/README.md

# Custom Azure Logs

The Custom Azure Logs integration collects logs from Azure Event Hub.

Use the integration to collect logs from:

> * Azure services that support exporting logs to Event Hub

Reviewer: nit:
```suggestion
* Azure services that support exporting logs to an Event Hub.
```

> * Any other source that can send logs to an Event Hub

Reviewer: ```suggestion
* Any other source that can send logs to an Event Hub.
```

> ## Data streams

Reviewer: Just checked the other input packages and those are not using the datastreams. Do you think the data stream section is relevant for the input package?

Author: I think it's still relevant.

We should inform users that the input package creates an index template matching their chosen dataset. This section will differ from non-input integrations, where integration developers define the data streams and datasets.

After rereading this section, I will change it a little to reflect this difference better.

Author: Updated.

The Custom Azure Logs Input integration collects one types of data streams: logs.

> The integration does not comes with a pre-defined data stream. You can select your dataset and namespace of choice when configuring the integration.

Reviewer 2: ```suggestion
The integration does not use a pre-defined Elastic data stream. You can select your dataset and namespace of choice when configuring the integration.
```

Author: Thanks for the fixes! ❤️ 

Note: `elastic-package` generates the file `packages/azure_logs/docs/README.md` using the content from the file `packages/azure_logs/_dev/build/docs/README.md`.

When you suggest docs fixes (VERY welcome!), please apply them to the files in the `_dev/build/docs` folder 🙇

Reviewer 2: ohh yes, good point 👍   will make sure to remember this next time :)

For example, if you select `azure.custom` as your dataset, and `default` as your namespace, the integration will send the data to the `logs-azure.custom-default` data stream.

Custom Logs integrations give you all the flexibility you need to configure the integration to your needs.

## Requirements

You need Elasticsearch for storing and searching your data and Kibana for visualizing and managing it.
You can use our hosted Elasticsearch Service on Elastic Cloud, which is recommended, or self-manage the Elastic Stack on your own hardware.

Before using the Custom Azure Logs you will need:

* One **event hub** to store in-flight logs exported by Azure services (or other sources) and make them available to Elastic Agent.
* A **storage account** to store information about logs consumed by the Elastic Agent.

### Event Hub

> [Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-about) is a data streaming platform and event ingestion service. It can receive and temporary store millions of events.

Reviewer: ```suggestion
[Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-about) is a data streaming platform and event ingestion service. It can receive and process millions of events in real-time.
```

Elastic Agent with the Custom Azure Logs integration will consume logs from the Event Hubs service.

```text
  ┌────────────────┐      ┌───────────┐
  │   myeventhub   │      │  Elastic  │
  │ <<Event Hub>>  │─────▶│   Agent   │
  └────────────────┘      └───────────┘
```

To learn more about Event Hubs, refer to [Features and terminology in Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-features).

### Storage Account Container

The [Storage account](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-overview) is a versatile Azure service that allows you to store data in various storage types, including blobs, file shares, queues, tables, and disks.

> The Custom Azure Logs integration requires a Storage account container to work.

Reviewer: Can we make the `storage account` case consistent across different sections?

Author: Updated!

Consumer groups allow multiple Elastic Agents assigned to the same agent policy to work together; this enables horizontal scaling of the logs processing when required.

In most cases, you can use the default consumer group named `$Default`. If `$Default` is already used by other applications, you can create a consumer group dedicated to the Azure Logs integration.

#### Connection string

The Elastic Agent requries a connection string to access the event hub and fetch the exported logs. The connection string contains details about the event hub used and the credentials required to access it.

To get the connection string for your Event Hubs namespace:

1. Visit the **Event Hubs namespace** you created in a previous step.
1. Select **Settings** > **Shared access policies**.

Create a new Shared Access Policy (SAS):

1. Select **Add** to open the creation panel.
1. Add a **Policy name** (for example, "ElasticAgent").
1. Select the **Listen** claim.
1. Select **Create**.

When the SAS Policy is ready, select it to display the information panel.

> Take note of the **Connection string–primary key**, which you will use later when specifying a **connection_string** in the integration settings.

Reviewer: The separator for `connection string-primary key`. Is this a typo?

Author: Oh no, as weird as it looks, it's the actual field label on the Azure Portal:

![CleanShot 2024-11-05 at 17 47 55@2x](https://github.com/user-attachments/assets/66c884d3-e246-4692-9d11-80585dccb1e4)

# Custom Azure Logs

The Custom Azure Logs integration collects logs from Azure Event Hub.

Use the integration to collect logs from:

* Azure services that support exporting logs to Event Hub
* Any other source that can send logs to an Event Hub

## Data streams

The Custom Azure Logs integration only supports logs data streams.

This custom integration does not use a pre-defined Elastic data stream like standard integrations do (for example, `logs-azure.activitylogs-default` for Activity logs). You can take control and build your own data stream by selecting your dataset and namespace of choice when configuring the integration.

For example, if you select `mydataset` as your dataset, and `default` as your namespace, the integration will send the data to the `logs-mydataset-default` data stream.

The integration sets up a dedicated index template named `logs-mydataset` with the `logs-mydataset-*` index pattern. You can then customize it using a custom pipeline and custom mappings.

> Custom Logs integrations give you all the flexibility you need to configure the integration to your needs.

Author: No, in this case, I am referring to integration or input packages designed to collect custom logs. With "custom logs", I mean any log format (usually JSON document) that may come from any source.

Maybe I should focus on the Custom Azure Logs integration only and change this phrase into:

> The Custom Azure Logs integration gives you all the flexibility you need to collect, parse, and map log events to your needs.

WDYT?

> Custom Logs integrations give you all the flexibility you need to configure the integration to your needs.

Reviewer 3: definitely better, it avoid ambiguity 👍

For example, if the integration is configured to fetch data from an event hub with four partitions, the Agent will create the following:

* One Storage Account container.
* Four blobs in that container.

The information stored in the blobs is small (usually < 500 bytes per blob) and accessed frequently. Elastic recommends using the Hot storage tier.

You need to keep the Storage Account container as long as you need to run the integration with the Elastic Agent. If you delete a Storage Account container, the Elastic Agent will stop working and create a new one the next time it starts.

By deleting a Storage Account container, the Elastic Agent will lose track of the last message processed and start processing messages from the beginning of the event hub retention period.

## Setup

Before adding the integration, you must complete the following tasks.

### Create an Event Hub

The event hub receives the logs exported from the Azure service and makes them available for the Elastic Agent to read.

Here's a high-level overview of the required steps:

* Create a resource group, or select an existing one.
* Create an Event Hubs namespace.

> * Create an event hub.

Author: I know it looks weird.

I picked up this capitalization from:

- Event Hubs: it's the actual _service_ name
- event hub: it's the actual resource that receives the messages 
- namespace: the cluster that holds one or more event hubs is called "namespace"

The text sometimes refers to the service name ("Event Hubs") and sometimes to the message recipient ("event hub").

Azure picked an ambiguous naming scheme for the Event Hubs service.

Here's the equivalent to the database context:

| Event Hubs | Databases |
|--------|--------|
| Event Hubs | Oracle Database |
| namespace | server |
| event hub | table | 

Microsoft seems to apply the same rules in their docs. For example, the doc [Features and terminology in Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-features) apply these rules consistently.

> * Create an event hub.

Author: I pushed an update to fix all inconsistencies with the naming used at https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-features

Let me know if this makes sense or if we want to take a different approach.

> * Create an event hub.

Reviewer 3: In the name of consistency, let's follow what is applied in the Microsoft docs.

Use the following settings to configure the Azure Logs integration when you add it to Fleet.

`eventhub` :
_string_
A fully managed, real-time data ingestion service. Elastic recommends using only letters, numbers, and the hyphen (-) character for event hub names to maximize compatibility. You can use existing event hubs having underscores (_) in the event hub name; in this case, the integration will replace underscores with hyphens (-) when it uses the event hub name to create dependent Azure resources behind the scenes (e.g., the Storage Account container to store event hub consumer offsets). Elastic also recommends using a separate event hub for each log type as the field mappings of each log type differ.
Default value `insights-operational-logs`.

`consumer_group` :
_string_
Enable the publish/subscribe mechanism of Event Hubs with consumer groups. A consumer group is a view (state, position, or offset) of an entire event hub. Consumer groups enable multiple consuming applications to each have a separate view of the event stream, and to read the stream independently at their own pace and with their own offsets.
Default value: `$Default`

`connection_string` :
_string_

> The connection string is required to communicate with Event Hubs. See [Get an Event Hubs connection string](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string) for more information.

Reviewer 3: ```suggestion
The connection string is required to communicate with Event Hubs. Check [Get an Event Hubs connection string](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string) for more information.
```

A Blob Storage Account is required to store/retrieve/update the checkpoint information of the event hub messages. This allows the integration to resume processing messages left when the user stops it.

`storage_account` :
_string_
The name of the Storage Account that stores the checkpoint information.
`storage_account_key` :
_string_
The Storage Account key. Key to authorize access to data in your Storage Account.

`storage_account_container` :
_string_

> The Storage Account container is where the integration stores the checkpoint data for the consumer group. It is an advanced option to use with extreme care. You MUST use a dedicated Storage Account container for each Azure log type (activity, sign-in, audit logs, and others). DO NOT REUSE the same container name for more than one Azure log type. See [Container Names](https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names) for details on naming rules from Microsoft. The integration generates a default container name if not specified.

Reviewer 3: ```suggestion
The Storage Account container is where the integration stores the checkpoint data for the consumer group. It is an advanced option to use with extreme care. You MUST use a dedicated Storage Account container for each Azure log type (activity, sign-in, audit logs, and others). DO NOT REUSE the same container name for more than one Azure log type. Check [Container Names](https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names) for details on naming rules from Microsoft. The integration generates a default container name if not specified.
```

`pipeline` :
_string_
Optional. Overrides the default ingest pipeline for this integration.

`resource_manager_endpoint` :
_string_

> Optional. By default, the integration uses the Azure public environment. To override this and use a different Azure environment, users can provide a specific resource manager endpoint

Reviewer 3: ```suggestion
Optional. By default, the integration uses the Azure public environment. To override this and use a different Azure environment, users can provide a specific resource manager endpoint.
```
